config/initializers/rack_attack.rb

```diff
@@ -81,25 +85,33 @@ def self.protected_route?(protected_actions, path, method)
   # throttle logins for another user and force their login requests to be
   # denied, but that's not very common and shouldn't happen to you. (Knock
   # on wood!)
-  throttle("logins/handler", limit: REQUEST_LIMIT, period: LIMIT_PERIOD) do |req|
-    if req.path == "/session" && req.post?
-      # return the handler if present, nil otherwise
-      req.params['session']['who'].presence if req.params['session']
+  protected_sessions_action = [{ controller: "sessions", action: "create" }]
+
+  throttle("logins/handle", limit: REQUEST_LIMIT, period: LIMIT_PERIOD) do |req|
+    req.params['session']['who'].presence if protected_route?(protected_sessions_action, req.path, req.request_method) && req.params['session']
+  end
+
+  throttle("api_key/basic_auth", limit: REQUEST_LIMIT, period: LIMIT_PERIOD) do |req|
+    if protected_route?(protected_api_key_action, req.path, req.request_method)
+      action_dispatch_req = ActionDispatch::Request.new(req.env)
+      ActionController::HttpAuthentication::Basic.user_name_and_password(action_dispatch_req).first.presence
```

Comment:
As above, over-throttling some accounts (only those with colliding normalized handles) is preferable to a bypass vulnerability. Normalize the username here too:

```rb
who = ActionController::HttpAuthentication::Basic.user_name_and_password(action_dispatch_req).first
User.normalize_email(who).presence
```

I suspect the `.presence` call should be the last method call in the chain, so blank strings are treated as `nil` and do not persist a throttle in memcached.